filters/auth/grantlogout.go

```diff
@@ -0,0 +1,202 @@
+//
+// grantLogout filter
+//
+// Used for revoking access/refresh tokens generated by the oauthGrant() filter.
+//
+
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	GrantLogoutName           = "grantLogout"
+	revokeTokenKey            = "token"
+	revokeTokenTypeKey        = "token_type_hint"
+	refreshTokenType          = "refresh_token"
+	accessTokenType           = "access_token"
+	unsupportedTokenTypeError = "unsupported_token_type"
+)
+
+type grantLogoutSpec struct {
+	config OAuthConfig
+}
+
+type grantLogoutFilter struct {
+	config OAuthConfig
+}
+
+type revokeErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+type basicAuthCredentials struct {
+	clientId     string
+	clientSecret string
+}
+
+func (*grantLogoutSpec) Name() string { return GrantLogoutName }
+
+func (s *grantLogoutSpec) CreateFilter([]interface{}) (filters.Filter, error) {
+	return &grantLogoutFilter{
+		config: s.config,
+	}, nil
+}
+
+func (f *grantLogoutFilter) getBasicAuthCredentials() (*basicAuthCredentials, error) {
+	clientID := f.config.GetClientID()
+	if clientID == "" {
+		return nil, errors.New("failed to create token revoke auth header: no client ID")
+	}
+
+	clientSecret := f.config.GetClientSecret()
+	if clientSecret == "" {
+		return nil, errors.New("failed to create token revoke auth header: no client secret")
+	}
+
+	return &basicAuthCredentials{
+		clientId:     clientID,
+		clientSecret: clientSecret,
+	}, nil
+}
+
+func getErrorResponse(response *http.Response) *revokeErrorResponse {
+	buf, err := ioutil.ReadAll(response.Body)
+	if err == nil && buf != nil {
+		var errorResponse revokeErrorResponse
+		err = json.Unmarshal(buf, &errorResponse)
+
+		if err == nil {
+			return &errorResponse
+		}
+	}
+
+	return nil
+}
+
+func responseToError(response *http.Response, tokenType string) error {
+	errorResponse := getErrorResponse(response)
```

Comment:
`getErrorResponse` does not check response status. If response does not contain proper `revokeErrorResponse` json the next line would :boom:  

Maybe try to inline `getErrorResponse` to get straightforward flow. With that it could bail out early, e.g. on json parse error:
```go
err = json.Unmarshal(buf, &errorResponse)
if err != nil {
    return err
}
```

Reply:
I see that it gets here only on 400, though we can not rely that 400 would always return proper json.

Reply:
This is fixed with the new flow where we made sure that the body always gets read out.